apis/core/v1alpha1/composition.go

```diff
@@ -0,0 +1,164 @@
+// transformers.
+type CompositionPatch struct {
+
+	// FromFieldPath is the path of the field on the upstream resource whose value
+	// to be used as input.
+	FromFieldPath string `json:"fromFieldPath"`
+
+	// ToFieldPath is the path of the field on the base resource whose value will
+	// be changed with the result of transforms.
+	ToFieldPath string `json:"toFieldPath,omitempty"`
+
+	// Transforms are the list of functions that are used as a FIFO pipe for the
+	// input to be transformed.
+	Transforms []Transform `json:"transforms,omitempty"`
+}
+
+// Patch runs transformers and patches the target resource.
+func (c *CompositionPatch) Patch(base, target *fieldpath.Paved) error {
+	in, err := base.GetValue(c.FromFieldPath)
+	if err != nil {
+		return err
+	}
+	out := in
+	for i, f := range c.Transforms {
+		out, err = f.Transform(out)
+		if err != nil {
+			return errors.Wrap(err, errTransformAtIndex(i))
+		}
+	}
+	return target.SetValue(c.ToFieldPath, out)
+}
+
+// TODO(muvaf): Reconsider the usefulness of Transformer interface. Nothing
+// implements it outside the package and its use helps only in Transform function
+// since actual Transformers are strong-typed with jsontags anyway.
+
+// Transformers resolve arbitrary input type to arbitrary output type. The
+// reasoning for this loose typing is that a Transformer may have an input of type
+// A but output of type B; given that there will be many pairs like this, it makes
+// more sense to enforce types at the lowest level of the chain which is content
+// of actual Resolve functions of individual transformers.
+//type Transformer interface {
+//
+//	// Resolve runs the logic of Transformer to produce an output from the input.
+//	Resolve(input interface{}) (interface{}, error)
+//}
+
+// Transform is a unit of process whose input is transformed into an output with
+// the supplied configuration.
+type Transform struct {
+
+	// Type of the transform to be run.
+	Type string `json:"type"`
+
+	// Math is used to transform input via mathematical operations such as multiplication.
+	Math *MathTransform `json:"math,omitempty"`
+
+	// Map uses input as key in the given map and returns the value.
+	Map *MapTransform `json:"map,omitempty"`
+}
+
+// Transform calls the appropriate Transformer.
+func (t *Transform) Transform(input interface{}) (interface{}, error) {
+	var transformer interface {
+		Resolve(input interface{}) (interface{}, error)
+	}
+	switch t.Type {
+	case "math":
+		transformer = t.Math
+	case "map":
+		transformer = t.Map
+	default:
+		return 0, errors.New(errTypeNotSupported(t.Type))
+	}
+	if transformer == nil {
+		return nil, errors.New(errConfigMissing(t.Type))
+	}
+	out, err := transformer.Resolve(input)
+	return out, errors.Wrap(err, errTransformWithType(t.Type))
+}
+
+// MathTransform conducts mathematical operations on the input with the given
+// configuration in its properties.
+type MathTransform struct {
```

Comment:
I wonder if the input should be an algebraic expression like `x * 1024` or `(x / 1024) * 1000` as string and the controller parse it using something like [govaluate](https://github.com/Knetic/govaluate).

Reply:
Hrm, sounds like that would get us a lot of power pretty easily. Do you think the UX would be easier for folks to write the YAML if we used govaluate?

Reply:
It seems like there's a bit of a tradeoff here. It's more work to model each potential operation in YAML, but we can validate it and potentially limit what folks can do. On the other hand usnig. govaluate would cover a lot of use cases fast.

Reply:
I imagine most of the operations will be simple ones like `(x / 1024) * 1000`. The most common use case I can think of this math operation is conversions between `GiB` and `GB` and `MB` etc.

> It's more work to model each potential operation in YAML, but we can validate it and potentially limit what folks can do.

For the conversion use-case, it's probably easy to just multiply with `1000/1024` -> `0,9765625`, so people can take shortcuts like that and use `multiply` only instead of stacking division and multiplication transforms on top of each other. But for sure, it's easier to see what's happening if it's like `(x / 1024) * 1000`, though I have to say this will be new; I haven't stumbled upon a case where the input is a formula.

Reply:
Well, actually they might have to stack multiple transforms for simple conversion: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types